extensions/eclipse-che-theia-plugin-remote/src/node/plugin-remote-node-impl.ts

```diff
@@ -0,0 +1,253 @@
+        const rawModel = plugin.rawModel;
+        await this.pluginRemoteBrowser.$definePluginPackage(pluginId, rawModel);
+
+        if (activatedPlugin && activatedPlugin.exports) {
+
+            // do we have prototype ?
+            const exported = activatedPlugin.exports;
+
+            const prototype = Object.getPrototypeOf(exported);
+            let proxyNames: Array<string> = [];
+            if (prototype) {
+                proxyNames = proxyNames.concat(Object.getOwnPropertyNames(prototype));
+            }
+
+            // ok now need to send that back only if there are some exports
+            if (proxyNames.length > 0) {
+                await this.pluginRemoteBrowser.$definePluginExports(pluginId, proxyNames);
+            }
+        }
+    }
+
+    async $activatePlugin(pluginId: string): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const pluginManagerInternal = (this.pluginManager as any);
+        await pluginManagerInternal.activatePlugin(pluginId);
+    }
+
+    async $definePluginPackage(pluginId: string, pluginPackage: PluginPackage): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const plugin = (this.pluginManager as any).registry.get(pluginId);
+        if (plugin) {
+            plugin.rawModel = pluginPackage;
+        }
+    }
+
+    // tslint:disable-next-line: no-any
+    deasyncPromise(promise: Promise<any>) {
+        // tslint:disable-next-line: no-any
+        let result: any;
+        // tslint:disable-next-line: no-any
+        let error: any;
+        let done = false;
+        promise.then(res => {
+            result = res;
+        }, err => {
+            error = err;
+        }).then(() => {
+            done = true;
+        });
+        deasync.loopWhile(() => !done);
+        if (error) {
+            throw error;
+        }
+        return result;
+    }
+
+    async $definePluginExports(hostId: string, pluginId: string, proxyNames: string[]): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const pluginManagerInternal = (this.pluginManager as any);
+
+        // add into the activatedPlugins stuff
+        // tslint:disable-next-line: no-any
+        const activatedPlugin: any = {};
+        // tslint:disable-next-line: no-any
+        const proxyExports: any = {};
+        activatedPlugin.exports = proxyExports;
+        const remoteBrowser = this.pluginRemoteBrowser;
+        const deasyncPromise = this.deasyncPromise;
+        // add proxy
+        const callId = this.callId++;
+        proxyNames.forEach(entryName => {
+            console.log(`Proxyfing exports ${hostId} ${pluginId}/${entryName}`);
+            // tslint:disable-next-line: no-any
+            proxyExports[entryName] = (...args: any[]) => {
+                // need to keep arguments only if arguments have functions as functions can't be propagated remotely
+                const hasFunctions = args.some(arg => typeof arg === 'function');
```

Comment:
There are frameworks that we can use for remoting like https://github.com/stanford-oval/transparent-rpc It even handles function call on the objects you pass as arguments.

Reply:
This framework doesn't do more than the RPC framework we already have on plugins. (BTW we have multi-plexing in theia RPC)
And functions are already handled by the underlying RPC protocol.

The real issue is for callbacks parameters that should not be sent: A function that returns a function